prov/gni/src/gnix_msg.c

```diff
@@ -1680,3 +2529,434 @@ ssize_t _gnix_send(struct gnix_fid_ep *ep, uint64_t loc_addr, size_t len,
 	}
 	return ret;
 }
+
+ssize_t _gnix_recvv(struct gnix_fid_ep *ep, const struct iovec *iov,
+		    void **desc, size_t count, uint64_t src_addr, void *context,
+		    uint64_t flags, uint64_t ignore, uint64_t tag)
+{
+	int i, ret = FI_SUCCESS;
+	size_t cum_len = 0;
+	struct gnix_fab_req *req = NULL;
+	struct gnix_address gnix_addr;
+	fastlock_t *queue_lock = NULL;
+	struct gnix_tag_storage *posted_queue = NULL;
+	struct gnix_tag_storage *unexp_queue = NULL;
+	uint64_t match_flags;
+	int tagged = flags & FI_TAGGED;
+	struct fid_mr *auto_mr[count];
+
+	if (!ep->recv_cq) {
+		return -FI_ENOCQ;
+	}
+
+	if (!tagged) {
+		if (!ep->ep_ops.msg_send_allowed)
+			return -FI_EOPNOTSUPP;
+
+		tag = 0;
+		ignore = ~0;
+	} else {
+		if (!ep->ep_ops.tagged_send_allowed)
+			return -FI_EOPNOTSUPP;
+	}
+
+	match_flags = flags & (FI_CLAIM | FI_DISCARD | FI_PEEK);
+
+	/*
+	 * Lookup the gni addr in the av_table or av_hashtable.
+	 * If the gni addr doesn't exist the addr is FI_ADDR_UNSPEC,
+	 * meaning this remote node wants to receive from all senders?
+	 */
+	ret = __gnix_msg_addr_lookup(ep, src_addr, &gnix_addr);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	/* calculate cumulative size of the iovec buf lens */
+	for (i = 0; i < count; i++) {
+		cum_len += iov[i].iov_len;
+	}
+
+	/*
+	 * Initialize the tag storage objects.
+	 * The posted_queue holds information about receives that have
+	 * been posted on the remote endpoint.
+	 *
+	 * The unexp_queue holds information about data that has arrived
+	 * prior to posting a receive on the remote endpoint.
+	 *
+	 * Both {unexp,posted}_queue objects have two sets, one for tagged
+	 * messages and the other for untagged messages.
+	 *
+	 * The untagged queues match based off the source address?
+	 *
+	 * The tagged queues match based off the tag and source address?
+	 *
+	 * A "message" is added to the unexpected queue when it arrives at a
+	 * remote endpoint and the completer_fn doesn't find an existing request
+	 * in the posted queue (i.e. no fi_recvs have been called (or posted)
+	 * on the remote endpoint).
+	 */
+	__gnix_msg_queues(ep, tagged, &queue_lock, &posted_queue, &unexp_queue);
+
+	COND_ACQUIRE(ep->requires_lock, queue_lock);
+
+	/*
+	 * Posting a recv, look for an existing request in the
+	 * unexpected queue.
+	 */
+retry_match:
+	req = _gnix_match_tag(unexp_queue, tag, ignore,
+			      match_flags, context, &gnix_addr);
+
+	if (req) {
+		GNIX_DEBUG(FI_LOG_EP_DATA, "UNEXPECTED, req = %p\n", req);
+		/* Found a matching request in the unexpected queue. */
+
+		/*
+		 * reset ep, it might be different than the ep the message came
+		 * in on.
+		 */
```

Comment:
> How can this happen?

Is it correct that, since there could be multiple eps for a given NIC, this unexpected message could've come in on an ep that is different than the ep we are posting this receive for?